libs/nio/src/main/java/org/elasticsearch/nio/NioSelector.java

```diff
@@ -153,13 +154,20 @@ void singleLoop() {
             preSelect();
             long nanosUntilNextTask = taskScheduler.nanosUntilNextTask(System.nanoTime());
             int ready;
-            if (nanosUntilNextTask == 0) {
-                ready = selector.selectNow();
-            } else {
-                long millisUntilNextTask = TimeUnit.NANOSECONDS.toMillis(nanosUntilNextTask);
-                // Only select until the next task needs to be run. Do not select with a value of 0 because
-                // that blocks without a timeout.
-                ready = selector.select(Math.min(300, Math.max(millisUntilNextTask, 1)));
+            try {
+                if (wokenUp.getAndSet(false) || nanosUntilNextTask == 0) {
+                    ready = selector.selectNow();
+                } else {
+                    long millisUntilNextTask = TimeUnit.NANOSECONDS.toMillis(nanosUntilNextTask);
+                    // Only select until the next task needs to be run. Do not select with a value of 0 because
+                    // that blocks without a timeout.
+                    ready = selector.select(Math.min(300, Math.max(millisUntilNextTask, 1)));
+
+                }
+            } finally {
+                if (wokenUp.get()) {
```

Comment:
Obviously this is confusing. And netty has that super-long comment explaining why. But I don't understand why we cannot do something like this and avoid it:

```
state is atomic integer

NOT_SELECTING = 0;
WOKEN_UP = 1;
SELECTING = 2;

wakeup-logic:

while(true) {
	int state = state.get();
	case state:
		NOT_SELECTING:
			if (state.compareAndSet(state, WOKEN_UP)) {
				return;
			}
		WOKEN_UP:
			return;
		SELECTING:
			if (state.compareAndSet(state, WOKEN_UP)) {
				wakeup();
				return;
			}
}

select logic:

if (wokenUp.getAndSet(SELECTING) == WOKEN_UP || nanosUntilNextTask == 0) {
	selectNow()
} else {
	select(300);
}
wokenUp.set(NOT_SELECTING);
```

We could also retain the atomic boolean and just set wokenUp=false immediately after finishing select. And I think that resolves the issue the netty comment outlines? The atomic boolean approach leads to a maximum of one wakeup per loop. Whereas the 3-state approach could lead to less wakeups.

Reply:
> We could also retain the atomic boolean and just set wokenUp=false immediately after finishing select. And I think that resolves the issue the netty comment outlines?

Even if you set it to `false` right after selecting, you will get redundant wakeups (you just cleared the wakeup state on the boolean, without actually running a `select`).

As far as the thee state solution vs. the Netty style solution I'm not sure:

The downside to the thee state solution seems that it runs into this:

1. IO thread is in `SELECTING` in `int state = state.get();`
2. IO thread selects 
3. IO thread is still in `SELECTING` in `if (state.compareAndSet(state, WOKEN_UP)) {` but done selecting but hasn't cleared the `interrupt` flag on the selector yet 
4. `WOKEN_UP` is set and wakeup is invoked
5. IO thread sets `NOT_SELECTING` 

-> Now the next thread that comes along will wake up redundantly one additional time (which could be at a time where it actually forces the wakeup syscall/fd write,  whereas the Netty style approach does that same potentially random wakeup on the IO thread when it wouldn't actually trigger a wakeup? Plus, we get less contention on the selector's internal lock in case of separate non-io threads invoking those two wake-ups.

Also, the code of the Netty style solution is much, much simpler => I'd prefer the simpler solution in this PR to the three state solution unless there's a good argument in its favor performance wise?

Reply:
> Also, the code of the Netty style solution is much, much simpler

I think my concern is that a change which uses an API (`wakeup`) clearly designed to be called by a different (notifying) thread in order to preemptively set an interrupt bit for the next time through the loop is not really simple. It requires a !4-5! paragraph comment in netty to explain. 

I agree the three state solution is also complex.

Let's consider this approach.

```java
            if (wokenUp.get() || nanosUntilNextTask == 0) {
                ready = selector.selectNow();
            } else {
                long millisUntilNextTask = TimeUnit.NANOSECONDS.toMillis(nanosUntilNextTask);
                // Only select until the next task needs to be run. Do not select with a value of 0 because
                // that blocks without a timeout.
                ready = selector.select(Math.min(300, Math.max(millisUntilNextTask, 1)));
            }
            wokenUp.set(false);
```

> Even if you set it to false right after selecting, you will get redundant wakeups

I'm not really sure I understand this concern.

If we're talking about:

T1 calls wakeup - takes interrupt lock and writes to FD
Event loop thread sets `wokenUp = false`
T2 calls wakeup - takes interrupt lock and writes to FD

And the downside of that is two lock acquires and FD writes per select loop? That can happen anywhere `wokenUp = false` is called (before or after select). And as the netty comment outlines, that extra wakeup can lead to an interrupted `select()` call the next time through the loop. As outlined in scenario 2. And I agree with the comment that seems OK.

```
                        // 'wakenUp' is set to true too early if:
                        // 1) Selector is waken up between 'wakenUp.set(false)' and
                        //    'selector.select(...)'. (BAD)
                        // 2) Selector is waken up between 'selector.select(...)' and
                        //    'if (wakenUp.get()) { ... }'. (OK)
                        //
                        // In the first case, 'wakenUp' is set to true and the
                        // following 'selector.select(...)' will wake up immediately.
                        // Until 'wakenUp' is set to false again in the next round,
                        // 'wakenUp.compareAndSet(false, true)' will fail, and therefore
                        // any attempt to wake up the Selector will fail, too, causing
                        // the following 'selector.select(...)' call to block
                        // unnecessarily.
                        //
                        // To fix this problem, we wake up the selector again if wakenUp
                        // is true immediately after selector.select(...).
                        // It is inefficient in that it wakes up the selector for both
                        // the first case (BAD - wake-up required) and the second case
                        // (OK - no wake-up required).
```

Setting `wokeUp=false` after the select seems to have similar behavior as netty's solution in scenario 2. But still solves scenario 1. And imo avoids a very counter intuitive usage of the `wakeup` API.

Reply:
Also I don't really even think the netty scenario 1 applies in our case. If the wakeup happens after wokeUp=false, wokeUp will = true the next time through the loop and we will use the non-blocking `selectNow()`.

Reply:
> Also I don't really even think the netty scenario 1 applies in our case. If the wakeup happens after wokeUp=false, wokeUp will == true the next time through the loop and we will use the non-blocking selectNow().

I think this is the important bit here, thanks for noticing! I wonder with that being the case ... do we even need any guards after the `select` logic? Can't we just remove 

```java
                if (wokenUp.get()) {
                    selector.wakeup();
                }
```

and be happy?

Reply:
Yeah. I think removing that would be sufficient since we use the non-blocking selectNow().

Reply:
:) removed it